src/waveform/renderers/qopengl/waveformrendererfiltered.cpp

```diff
@@ -0,0 +1,255 @@
+
+    float bandGain[3];
+    bandGain[0] = lowGain;
+    bandGain[1] = midGain;
+    bandGain[2] = highGain;
+
+    const float breadth = static_cast<float>(m_waveformRenderer->getBreadth()) * devicePixelRatio;
+    const float halfBreadth = breadth / 2.0f;
+
+    const float heightFactor = allGain * halfBreadth / 255.f;
+
+    // Effective visual index of x
+    double xVisualSampleIndex = firstVisualIndex;
+
+    const int numValuesPerLine = 12; // 2 values per vertex, 3 vertices per triangle, 2 triangles
+
+    int reserved[4];
+    // low, mid, high
+    for (int bandIndex = 0; bandIndex < 3; bandIndex++) {
+        m_lineValues[bandIndex].clear();
+        reserved[bandIndex] = numValuesPerLine * length;
+        m_lineValues[bandIndex].reserve(reserved[bandIndex]);
+    }
+
+    // the horizontal line
+    reserved[3] = numValuesPerLine;
+    m_lineValues[3].clear();
+    m_lineValues[3].reserve(reserved[3]);
+
+    addRectangle(m_lineValues[3],
+            0.f,
+            halfBreadth - 0.5f * devicePixelRatio,
+            static_cast<float>(length),
+            halfBreadth + 0.5f * devicePixelRatio);
+
+    for (int pos = 0; pos < length; ++pos) {
+        // Our current pixel (x) corresponds to a number of visual samples
+        // (visualSamplerPerPixel) in our waveform object. We take the max of
+        // all the data points on either side of xVisualSampleIndex within a
+        // window of 'maxSamplingRange' visual samples to measure the maximum
+        // data point contained by this pixel.
+        double maxSamplingRange = gain / 2.0;
+
+        // Since xVisualSampleIndex is in visual-samples (e.g. R,L,R,L) we want
+        // to check +/- maxSamplingRange frames, not samples. To do this, divide
+        // xVisualSampleIndex by 2. Since frames indices are integers, we round
+        // to the nearest integer by adding 0.5 before casting to int.
+        int visualFrameStart = int(xVisualSampleIndex / 2.0 - maxSamplingRange + 0.5);
+        int visualFrameStop = int(xVisualSampleIndex / 2.0 + maxSamplingRange + 0.5);
+        const int lastVisualFrame = dataSize / 2 - 1;
+
+        // We now know that some subset of [visualFrameStart, visualFrameStop]
+        // lies within the valid range of visual frames. Clamp
+        // visualFrameStart/Stop to within [0, lastVisualFrame].
+        visualFrameStart = math_clamp(visualFrameStart, 0, lastVisualFrame);
+        visualFrameStop = math_clamp(visualFrameStop, 0, lastVisualFrame);
+
+        int visualIndexStart = visualFrameStart * 2;
+        int visualIndexStop = visualFrameStop * 2;
+
+        visualIndexStart = std::max(visualIndexStart, 0);
+        visualIndexStop = std::min(visualIndexStop, dataSize);
+
+        // 3 bands, 2 channels
+        float max[3][2]{};
+
+        for (int i = visualIndexStart; i < visualIndexStop; i += 2) {
+            const WaveformData& waveformData = data[i];
+            const WaveformData& waveformDataNext = data[i + 1];
```

Comment:
I think this makes the following code more easier to understand.
```suggestion
            const WaveformData& waveFormSampleRightChannel = data[i];
            const WaveformData& waveFormSampleLeftChannel = data[i + 1];
```

Reply:
Sure. this is the terminology used in the legacy renderers, but I agree this is much more clear.

So I started doing this but than I realised I could get rid of a lot of the code duplication for the channels by using arrays with 2 elements and for (int chn=0;chn<2;chn++).

Reply:
which is done